torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,332 @@
+import torch
+from collections import defaultdict
+from torch._six import container_abcs
+
+
+class _PerDeviceHelper(object):
+    """
+    Helper class to serve copies of a master tensor that may be needed on multiple devices.
+    Intended to be used as a temporary.
+    """
+    def __init__(self, master_tensor):
+        self.master = master_tensor
+        self._per_device_tensors = {}
+
+    def get(self, device):
```

Comment:
Hmm. I think we have to do better on this implementation. You're cloning even when the tensor is already on the device? Is that really what you want to do?

Reply:
yes, as we discussed in slack